Update content trust docs to reflect latest notary compose file changes

cyli opened the pull request:
This updates the content trust sandbox docs to reflect how to test with the latest notary server.  It also updates, in the same doc, how to get the latest docker client binary that isn't based off of master (so that there isn't a client/server API mismatch).

This addresses some of the confusion from https://github.com/docker/notary/issues/762.

![cute picture](https://pbs.twimg.com/media/CReQaj6UsAACl9h.jpg:small)
Signed-off-by: cyli cyli@twistedmatrix.com

cyli committed:
Update content trust docs to reflect latest notary compose file changes, and to simplify
the instructions by providing a single compose file that runs the notary server, registry,
and a docker-in-docker trust sandbox.

Signed-off-by: cyli <cyli@twistedmatrix.com>

SvenDowideit commented:
LGTM

cyli commented:
Thanks!  The tests failed so I wasn't sure whether that was me.  I think I
may have a couple more changes though based on more feedback from a user. :)

On Thursday, June 2, 2016, Sven Dowideit <notifications@github.com> wrote:

> LGTM
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/23195#issuecomment-223468659>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AANn4G8kOUa_FMJsVuAdCKVlAa2XYi4Nks5qH4QPgaJpZM4IsOtc>
> .
>

SvenDowideit commented:
retest this please

SvenDowideit commented:
the test failure was a combination of my mistake, and a system failure rebuilding the hub images (GH's svn keeps failing)

SvenDowideit commented:
retest this please

thaJeztah commented:
@cyli is this ready to merge? I see you added WIP?

cyli commented:
@thaJetzah I'm trying to validate this right now, since it doesn't really work for Docker for Mac (because I can't get the `/etc/hosts` file to take effect for the xhyve vm), so just setting up a regular VM to do so.  Once I've made sure that's right, yes.

cyli commented:
I am getting a weird result when pulling after pushing a signed image - it seems like signing an image signs in a target `""`, with no hashes and a size of 0 for me.  :|  Despite the fact that it should be the "latest" tag.  Can anyone else replicate this?

riyazdf commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -130,13 +130,13 @@ So, you'll need an entry for both the servers in your local `/etc/hosts` file.
-          $ git clone -b trust-sandbox https://github.com/docker/notary.git
+          $ git clone https://github.com/docker/notary.git
 
 3. Clone the `distribution` project.
```
I think we can remove this step now

cyli replied:
Done, thanks for spotting this!  I think I had previously removed this but messed up a stash conflict resolution.

riyazdf commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -76,29 +76,31 @@ So, you'll need an entry for both the servers in your local `/etc/hosts` file.
+        $ mkdir sandboximage
+        $ cd sandboximage
 
 4. Create a filed called `Dockerfile` with your favorite editor.
```
nit: s/`filed`/`file`

cyli replied:
👍 fixed, thanks!

cyli commented:
Ok, I think this might be ready for final review.  Thanks @avaid96 and @riyazdf for running through these with me.  Also just wanted to check with @diogomonica to verify I hadn't gotten any of the tampering stuff wrong.

riyazdf commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -274,49 +241,59 @@ What happens when data is corrupted and you try to pull it when trust is
+        sandboxregistry:5000/test/trusttest   <none>              cc7629d1331a        11 months ago       5.025 MB
 
-7. Remove the `trusttest:latest` image.
+8. Remove the `trusttest:latest` image.
```
It might be worth clarifying that we're doing this to clear our local cache, to test that we won't download malicious data from the registry on `pull`?

cyli replied:
Good point, thanks!

diogomonica replied:
yes.

riyazdf commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -327,5 +304,25 @@ Now, that you have a full Docker content trust sandbox on your local system,
 feel free to play with it and see how it behaves. If you find any security
 issues with Docker, feel free to send us an email at <security@docker.com>.
 
+## Troubleshooting
```
this section is awesome, thank you so much for adding it 👍

riyazdf commented:
LGTM after small comment!  thank you for bringing these docs up to date :)

diogomonica commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -38,18 +40,18 @@ production trust environment, and requires these additional components:
 | Container       | Description                                                                                                                                 |
 |-----------------|---------------------------------------------------------------------------------------------------------------------------------------------|
-| notarysandbox  | A container with the latest version of Docker Engine and with some preconfigured certifications. This is your sandbox where you can use the `docker` client to test trust operations. |
+| trustsandbox  | A container with the latest version of Docker Engine and with some preconfigured certifications. This is your sandbox where you can use the `docker` client to test trust operations. |
```
what is a preconfigured certification? :/

diogomonica commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -61,146 +63,114 @@ work exclusively with the Docker Hub, you would not need with these components.
+        $ cd trustsandbox
 
-2. Change into your `notarysandbox` directory.
+2. Create a filed called `Dockerfile` with your favorite editor.  For example, using vim:
```
I think we should just prepackage this as an image, have all the instructions simple do `docker run docker/trustsandbox` and have an appendix of how to build your own trustsandbox image.

cyli replied:
That would be great actually.

cyli replied:
Do you have permission to push to the docker org?

diogomonica replied:
Yeah, I do, but you can ask @dmp42

diogomonica commented:
LGTM after nits.

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -23,12 +23,14 @@ These instructions assume you are running in Linux or Mac OS X. You can run
-This sandbox requires you to install two Docker tools: Docker Engine and Docker
-Compose. To install the Docker Engine, choose from the [list of supported
-platforms](../../installation/index.md). To install Docker Compose, see the
+This sandbox requires you to install two Docker tools: Docker Engine >= 1.1.0
```
Given that Docker 1.5.0 and below is no longer supported by Docker Hub, I think we should remove the version requirement here; all docs are for the current version here, wdyt?

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -61,146 +63,114 @@ work exclusively with the Docker Hub, you would not need with these components.
 
-        ENTRYPOINT ["bash"]
+        ENTRYPOINT ["dockerd-entrypoint.sh"]
+        CMD []
```
This `ENRYPOINT` and `CMD` is already defined in the `docker:dind` image, so are not needed here (see https://github.com/docker-library/docker/blob/866c3fbd87e8eeed524fdf19ba2d63288ad49cd2/1.11/dind/Dockerfile)

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -61,146 +63,114 @@ work exclusively with the Docker Hub, you would not need with these components.
-        $ sudo sh -c 'echo "127.0.0.1 sandboxregistry" >> /etc/hosts'
-
-
 ### Build the notarytest image
```
s/notarytest/notarysandbox/

thaJeztah commented on docs/security/trust/trust_sandbox.md:
````diff
@@ -61,146 +63,114 @@ work exclusively with the Docker Hub, you would not need with these components.
 ```
-$ docker run -it -v /var/run/docker.sock:/var/run/docker.sock --link notary_notaryserver_1:notaryserver --link sandboxregistry:sandboxregistry notarysandbox
-root@0710762bb59a:/#
+$ docker run -d --privileged --net notary_sig \
````
Wondering; can't we build and run the `trustsandbox` as part of the docker-compose file, then `docker-compose exec trustsandbox`

cyli replied:
It probably shouldn't go into the notary compose file, since the notary compose file is just for setting up the notary service.  Can we specify a compose file somewhere in the docker docs directory and have the users download and run that?

thaJeztah replied:
Yes, thinking what would make sense; I see the docker-compose file in the notary repository also builds notary from source; if that's something we want, it'd best be somewhere in the notary repository.

Perhaps having images available on Docker Hub would be good (I see no direct reason why people would have to build from source; especially since we don't tell them to checkout a specific tag/branch, so it's possible current master doesn't match this guide

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -61,146 +63,114 @@ work exclusively with the Docker Hub, you would not need with these components.
-root@0710762bb59a:/#
+$ docker run -d --privileged --net notary_sig \
+   --link sandboxregistry:sandboxregistry \
+   --link notary_server_1:notaryserver \
```
I don't think these links are required; docker-compose automatically adds aliases for the services

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -61,146 +63,114 @@ work exclusively with the Docker Hub, you would not need with these components.
-container.  When you destroy the container, you destroy the "play" keys.
+Running this container, based on the docker-in-docker image, lets you pull and push images while
+storing all the trusttest images, and trust keys and files inside the sandbox container.  When
+you destroy the container, you destroy the "play" keys and the cached trusttest images.
```
Actually, the `docker:dind` has a `VOLUME` for `/var/lib/docker`, so they are persisted in an anonymous volume. https://github.com/docker-library/docker/blob/866c3fbd87e8eeed524fdf19ba2d63288ad49cd2/1.11/dind/Dockerfile#L21

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -327,5 +308,25 @@ Now, that you have a full Docker content trust sandbox on your local system,
+
+    Get https://sandboxregistry:5000/v1/_ping: dial tcp: lookup sandboxregistry on 192.168.65.1:53: no such host
+
+Please make sure that your `trustsandbox` container is connected to the notary server and the registry:
```
s/PLease/Please/

or perhaps better; just remove "please"

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -327,5 +308,25 @@ Now, that you have a full Docker content trust sandbox on your local system,
+                "Links": [
+                    "/notary_server_1:/trustsandbox/notaryserver"
+                    "/sandboxregistry:/trustsandbox/sandboxregistry"
+                ],
```
See above notes w.r.t. links; I don't think you need the links (compose should set up the aliases if the V2 format is used)

thaJeztah commented:
@cyli thanks! Left some suggestions / notes; also, can you squash your commits?

cyli commented:
Thanks @thaJeztah!  I really like the suggestion of a compose file - I was thinking of doing the same earlier, but I didn't know where to put it.  If you think it'd be ok to add it to the docs directory, or some example docker directory, please let me know!

cyli commented:
@thaJeztah Thank you for all your suggestions!  I've produced a compose file which seems to work correctly for me, and inlined it in the docs.  Could you please take another look?

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -212,60 +150,57 @@ Now, you'll pull some images.
-        a9539b34a6ab: Image successfully pushed
-        b3dbab3810fc: Image successfully pushed
-        latest: digest: sha256:1d871dcb16805f0604f10d31260e79c22070b35abc71a3d1e7ee54f1042c8c7c size: 3348
+        / # docker push sandboxregistry:5000/test/trusttest:latest
```
Perhaps for this step, we should mention that they need to create passwords? (I realize the output mentions it, but it may be good to tell them upfront)

cyli replied:
Good idea

cyli replied:
Added below

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -274,49 +209,64 @@ What happens when data is corrupted and you try to pull it when trust is
+
+        aac0c133338d: Retrying in 5 seconds
+        a3ed95caeb02: Download complete
+        error pulling image configuration: unexpected EOF
```
(off topic) I this all the feedback you'll get? Think we should improve on that, I can imagine it being confusing that the only feedback is that docker keeps retrying, but no message *why*

cyli replied:
Yeah :|  Editing some of the other layers produces just the unexpected EOF message too, and not even the retrying message.  The message from the previous iteration of this doc was nicer.

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -21,186 +21,124 @@ overview](content_trust.md).
+the `trustsandbox` container, the Notary server, and the Registry server.
+
+
+1. Create a filed called `trustsandbox.yml` with your favorite editor.  For example, using vim:
```
I think we should create a regular `docker-compose.yml` file. That way they don't have to add the `-f trustsandbox.yml` when working with docker-compose. We should advice them to create a new directory first for the project ( e.g. `mkdir trustsandbox`)

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -21,186 +21,124 @@ overview](content_trust.md).
 [detailed instructions here](https://docs.docker.com/compose/install/).
 
-Finally, you'll need to have `git` installed on your local system or VM.
+Finally, you'll need to have `git` and a text editor installed on your local
```
`git` is probably no longer required now

thaJeztah commented:
Thanks @cyli, I tried the steps, and it works, just some minor suggestions to make it slightly easier even

cyli commented:
@thaJeztah Thank you for actually running through this and for your helpful feedback!  I think I've addressed your feedback, docwise anyway (the unhelpful EOF message is not addressed).

thaJeztah commented:
Thanks @cyli, I'll give it one more check, later tonight

> the unhelpful EOF message is not addressed

Should we open an issue for that to track it? I can imagine that (as a user), I'd be confused by "nothing happening"

cyli commented:
@thaJeztah Sounds good - opened https://github.com/docker/docker/issues/23493

thaJeztah commented on docs/security/trust/trust_sandbox.md:
```diff
@@ -328,4 +284,10 @@ feel free to play with it and see how it behaves. If you find any security
+anonymous volumes that have been created, just run the following command in the
+directory where you've created your Docker Compose file:
+
+        $ docker-compose -f trustsandbox.yml down -v
```
oh! you missed one; `-f trustsandbox.yml` can be removed here

cyli replied:
Good catch, thanks!

thaJeztah commented:
One final comment, but LGTM after that's fixed, thanks!

thaJeztah commented:
Thanks! Merging, because I think we've got enough discussion on this PR by now, and Sven already LGTM'd
